tutorials/haskell/servant-auth/tutorial.md

```diff
@@ -0,0 +1,566 @@
+
+-- | The home page does not do anything fancy, although just like all the
+-- other pages it features the menu that changes depending on whether the
+-- user is logged in or not.
+
+getHomeR :: Maybe Session -> App (AppView HomeView)
+getHomeR ms = mkAppView ms HomeView
+
+-- | The “Sign In” handler set the cookie and displayes a message.
+
+getSignInR :: App (Headers '[Header "set-cookie" ByteString] (AppView SignInView))
+getSignInR = do
+  let session = Session "mark" "mark@example.org"
+  AppContext {..} <- ask
+  mkAppView (Just session) (SignInView "You have signed in.")
+    >>= addSession appContextAuthSettings appContextRandomSource appContextServerKey session
+
+-- | Due to the use of the 'withSession' helper, this page is only available
+-- to logged in users.
+
+getMyPageR :: Maybe Session -> App (AppView MyPageView)
+getMyPageR ms = withSession ms $ \Session {..} ->
+  mkAppView ms (MyPageView sessionUsername sessionEmail)
+
+-- | The “Sign Out” page sets cookies to empty byte string destroyng the data.
+
+getSignOutR
+  :: Maybe Session
+  -> App (Headers '[Header "set-cookie" ByteString] (AppView SignOutView))
+getSignOutR ms = withSession ms $ \_ -> do
+  AppContext {..} <- ask
+  mkAppView Nothing (SignOutView "You have signed out.")
+    >>= addSession appContextAuthSettings appContextRandomSource appContextServerKey ()
+```
+
+As you can see the only interesting moment here in the use of `addSession`
+function that comes from the `servant-auth-cookie` package. Using the values
+from the `App` monad such as cookie settings and server key we encrypt the
+binary data of serialized `Session` and add it to response we are going to
+send to the client.
+
+For the sake of simplicity we don't show a “Sign In” form here, although it
+could be added trivially. Every time you request the “Sign In” page the app
+creates a cookie for the same user. Nothing stops you from performing a
+database lookup here and setting the real data.
+
+## Running the application
+
+To run the app we need to compile it, and execute the `servant-auth`
+executable (if you cloned our repository, you may have called the executable
+differently if you are recreating the application on your own):
+
+```bash
+$ stack build
+$ stack exec servant-auth
+```
+
+Then go to `localhost:8000` in your browser, you should see something like
+this:
+
+```
+TODO insert a screenshot here
+```
+
+Right now you are not logged in. To log in follow the “Sign In” link — you
+will notice that menu items have changed accordingly. Now you have the
+access to “My Page” and “Sign Out” endpoint that were not present on the
+menu previously and in fact were completely inaccessible for you.
+
+## Conclusion
+
+We have achieved our goal and created usable user interface staying close to
+Servant's philosophy. The current “generalized authentication” of Servant
+certainly can be used to craft any sort of authentication, although the
+feature still have strong “experimental” flavor in it.
```

Comment:
@mrkkrp 

- "a usable user interface"
- "although the feature still has a strong "experimental" flavor in it"